### ISSUE TITLE
Crash on Some Linux Distributions: “Cross‐device moves not yet implemented”

### ISSUE THREAD

**Issue by felix91gr** [SDGGiesbrecht/SDGCommandLine#35]
### Description

It fails. Below's how it fails.

### Demonstration

```console
felix@felix-X550LD ~/D/P/A/sdgCommandLine> swift test
Test Suite 'All tests' started at 2018-04-13 04:50:33.864
Test Suite 'debug.xctest' started at 2018-04-13 04:50:33.865
Test Suite 'InternalTests' started at 2018-04-13 04:50:33.865
Test Case 'InternalTests.testVersionSelection' started at 2018-04-13 04:50:33.865

$ swift --version
Swift version 4.1 (swift-4.1-RELEASE)
Target: x86_64-unknown-linux-gnu


$ swift package init --type executable
Creating executable package: tool
Creating Package.swift
Creating README.md
Creating .gitignore
Creating Sources/
Creating Sources/tool/main.swift
Creating Tests/


$ git version
git version 2.7.4

Expected Git 2.15.1. Attempting to continue anyway...

$ git init
Initialized empty Git repository in /tmp/ca.solideogloria.SDGCommandLine.Tests.debug/tool/.git/


$ git add .


$ git commit --m Initialized.
[master (root-commit) 48fcd9e] Initialized.
 4 files changed, 27 insertions(+)
 create mode 100644 .gitignore
 create mode 100644 Package.swift
 create mode 100644 README.md
 create mode 100644 Sources/tool/main.swift


$ git add .


$ git commit --m "Version 1.0.0"
[master 68839a4] Version 1.0.0
 1 file changed, 1 insertion(+), 1 deletion(-)


$ git tag 1.0.0


$ git clone "file:///tmp/ca.solideogloria.SDGCommandLine.Tests.debug/tool" "/tmp/ca.solideogloria.SDGCommandLine.Tests.debug/Package Clones/tool" --branch 1.0.0 --depth 1 --config advice.detachedHead=false
Cloning into '/tmp/ca.solideogloria.SDGCommandLine.Tests.debug/Package Clones/tool'...


$ swift build --configuration release --static-swift-stdlib
Compile Swift Module 'tool' (1 sources)
Linking ./.build/x86_64-unknown-linux/release/tool

Fatal error: Cross-device moves not yet implemented is not yet implemented: file Foundation/FileManager.swift, line 430
Exited with signal code 4

```

### Availability to Help

I can try to point to the causes if I can, and help by running tests, if needed.

### Solution/Design Thoughts

It might be related to #34. This one definitely fails because of [`FileManager.swift`'s incomplete implementation](https://github.com/apple/swift-corelibs-foundation/blob/master/Foundation/FileManager.swift). And to reference the same intention as in #34: they do [intend to finish `corelibs-foundation` at some point](https://forums.swift.org/t/shouldn-t-foundation-give-off-warnings-when-nsunimplemented-is-bound-to-appear/11576/8?u=felix91gr).

**Comment by felix91gr** [comment]
>  is there a chance these paths are somehow on distinct devices (external drives, remote desktop, etc?) the way you are running the tests?

No, not at all. This is on just one drive.

Something that was _very_ weird though, was that in order to be able to reproduce this failure after I'd fiddled with #34's test, I had to erase the folder created under `/tmp` by the test suite. Otherwise, the test would fail and the test suite wouldn't crash:

```console
felix@felix-X550LD ~/D/P/A/sdgCommandLine> swift test
Fetching https://github.com/SDGGiesbrecht/SDGCornerstone
Cloning https://github.com/SDGGiesbrecht/SDGCornerstone
Resolving https://github.com/SDGGiesbrecht/SDGCornerstone at 0.8.0
Compile Swift Module 'SDGLogic' (10 sources)
Compile Swift Module 'SDGControlFlow' (12 sources)
Compile Swift Module 'SDGConcurrency' (2 sources)
Compile Swift Module 'SDGMathematics' (30 sources)
Compile Swift Module 'SDGRandomization' (12 sources)
Compile Swift Module 'SDGGeometry' (5 sources)
Compile Swift Module 'SDGCollections' (35 sources)
Compile Swift Module 'SDGBinaryData' (6 sources)
Compile Swift Module 'SDGText' (28 sources)
Compile Swift Module 'SDGPersistence' (10 sources)
Compile Swift Module 'SDGLocalization' (28 sources)
Compile Swift Module 'SDGCornerstoneLocalizations' (3 sources)
Compile Swift Module 'SDGExternalProcess' (4 sources)
Compile Swift Module 'SDGPrecisionMathematics' (9 sources)
Compile Swift Module 'SDGCalendar' (43 sources)
Compile Swift Module 'SDGCornerstone' (1 sources)
Compile Swift Module 'SDGCommandLine' (29 sources)
Compile Swift Module 'SDGCommandLineTests' (12 sources)
Compile Swift Module 'SDGCommandLinePackageTests' (1 sources)
Linking ./.build/x86_64-unknown-linux/debug/SDGCommandLinePackageTests.xctest
Test Suite 'All tests' started at 2018-04-13 04:49:00.100
Test Suite 'debug.xctest' started at 2018-04-13 04:49:00.101
Test Suite 'InternalTests' started at 2018-04-13 04:49:00.101
Test Case 'InternalTests.testVersionSelection' started at 2018-04-13 04:49:00.101

$ swift --version
Swift version 4.1 (swift-4.1-RELEASE)
Target: x86_64-unknown-linux-gnu


$ swift package init --type executable
error: a manifest file already exists in this directory
Creating executable package: tool
/home/felix/Documents/Programming/AwesomeSystemSwift/sdgCommandLine/Tests/SDGCommandLineTests/All Tests/InternalTests.swift:206: error: InternalTests.testVersionSelection : failed - The operation could not be completed
Test Case 'InternalTests.testVersionSelection' failed (0.03 seconds)
Test Suite 'InternalTests' failed at 2018-04-13 04:49:00.131
	 Executed 1 test, with 1 failure (0 unexpected) in 0.03 (0.03) seconds
Test Suite 'debug.xctest' failed at 2018-04-13 04:49:00.131
	 Executed 1 test, with 1 failure (0 unexpected) in 0.03 (0.03) seconds
Test Suite 'All tests' failed at 2018-04-13 04:49:00.131
	 Executed 1 test, with 1 failure (0 unexpected) in 0.03 (0.03) seconds
```

> Thank you for reporting these, @felix91gr.

No problem! ^^

These following weeks are getting intense at my uni, so I might be apprearing only sporadically on Github because of well, studies. So if I'm not merging your PR super quickly, it's because my mind's somewhere else :P

**Comment by SDGGiesbrecht** [comment]
...but that is peripheral to the main issue.

**Comment by SDGGiesbrecht** [comment]
For comparison, a healthy test log (from master at 0.2.0) continues like [this](https://travis-ci.org/SDGGiesbrecht/SDGCommandLine/jobs/364809664#L1491-L1492).

**Comment by SDGGiesbrecht** [comment]
Ergo the problem occurs during one of the two move operations in [these lines](https://github.com/SDGGiesbrecht/SDGCommandLine/blob/bff590b072f9120ba4ae0db8b5f7b0e6e9cc3b87/Sources/SDGCommandLine/Package/Package.swift#L92-L106).

I will take another look tomorrow.

**Comment by SDGGiesbrecht** [comment]
@felix91gr,

Since my Linux machine does not encounter this error, I would need your help to debug it. (It is up to you how much effort you want to put into this.)

1. I have created a branch (troubleshoot35) that logs the origin and destination of each move. If you run the following (or something equivalent), **what are the failing paths? What do they actually contain on the drive after the crash? What are their permissions?, etc.**
```shell
git clone https://github.com/SDGGiesbrecht/SDGCommandLine
cd SDGCommandLine
git checkout troubleshoot35
swift test --filter SDGCommandLineTests.InternalTests/testVersionSelection
cd ..
rm -rf SDGCommandLine
```

2. Does the code fail in normal use, or just while testing? The following would execute the same code in a real use case. **Does it successfully print the usage help, or crash trying to retrieve the selected alternate version?**
```shell
git clone https://github.com/SDGGiesbrecht/Workspace
cd Workspace
# From HEAD of master:
swift run --configuration release workspace help •use‐version 0.6.0
cd ..
rm -rf Workspace
```

3. You can also try clearing all of [these paths](https://github.com/SDGGiesbrecht/SDGCornerstone/blob/26a5be3893f8327fa371cb4a540e669f73f7292c/Sources/SDGPersistence/FileManager.swift#L60-L65) of anything `ca.solideogloria...` related and trying again.

**Comment by felix91gr** [comment]
#### Results
I first followed the instructions of point (3). Then I:

* Ran (1),
* Followed (3) again to clear up the files, and 
* Ran (2)

This was the output of both runs:

1. [This](https://paste.ubuntu.com/p/YBZRWKSCnr/)
2. [This](https://paste.ubuntu.com/p/gvWR3y6fxS/)

**Comment by SDGGiesbrecht** [comment]
@felix91gr,

If `~/.cache` turns out to be a no‐go, the next question is where it is safe to put it instead.

To play around and find a place, you can use this quick sample project:

#### Set‐Up
```shell
mkdir testmove
cd testmove
swift package init --type executable
echo -e "// swift-tools-version:4.1\nimport PackageDescription\nlet package = Package(\n    name: \"testmove\",\n    dependencies: [\n        .package(url: \"https://github.com/SDGGiesbrecht/SDGCornerstone\", from: Version(0, 8, 0))\n    ],\n    targets: [\n        .target(name: \"testmove\", dependencies: [\"SDGPersistence\"])\n    ]\n)" > Package.swift
echo -e "import SDGPersistence\nProcessInfo.applicationIdentifier = \"com.felix91gr.testmove\"\ndo {\n    let temporaryDirectory = FileManager.default.url(in: .temporary, at: \"Directory\")\n    let temporaryFile = temporaryDirectory.appendingPathComponent(\"File\")\n    let cacheDirectory = FileManager.default.url(in: .cache, at: \"Directory\")\n\n    print(\"Creating temporary file...\")\n    try \"Hello, world"'!'"\".save(to: temporaryFile)\n    print(\"Creation succeeded.\")\n\n    try? FileManager.default.removeItem(at: cacheDirectory)\n    print(\"Moving to cache...\")\n    try FileManager.default.move(temporaryDirectory, to: cacheDirectory)\n    print(\"Move succeeded.\")\n} catch {\n    print(error)\n}" > Sources/testmove/main.swift
swift package edit SDGCornerstone
```

#### Execute Test
```shell
swift run testmove
```

#### Try Another Path
```shell
open Packages/SDGCornerstone/Sources/SDGPersistence/FileManager.swift
```
...then change line 63 (`path = NSHomeDirectory() + "/.cache"`).

#### Clean‐Up
```shell
cd ..
rm -rf testmove
```

#### Notes

> `cache`
> For caches that are saved to improve performance. The operating system may empty this if it needs to create space. Do not save anything here that cannot be regenerated if necessary.

[Earlier versions used `/var/cache`](https://github.com/SDGGiesbrecht/SDGCornerstone/commit/226e1a045495b516b3ef3da7ae000c69a192fc70#diff-e2fa74d64da7b19eb8016a559464e410L60) until it ran into permissions issues.

**Comment by felix91gr** [comment]
### Permissions

These are my permissions:

```console
felix@felix-X550LD ~> ls -ld ~/.cache
drwxrwxr-x 22 felix felix 4096 abr 14 00:55 /home/felix/.cache/
felix@felix-X550LD ~> ls -ld ~
drwx------ 46 felix felix 12288 abr 15 15:32 /home/felix/
```
### Testing directories

I tried both making `testmove` write to `~/.cache` and making it write to an arbitrary location `~/testdir`. This was the output for both:

```console
felix@felix-X550LD ~/D/P/A/testmove> swift run testmove
Compile Swift Module 'SDGPersistence' (10 sources)
Linking ./.build/x86_64-unknown-linux/debug/testmove
Creating temporary file...
Creation succeeded.
Moving to cache...
Fatal error: Cross-device moves not yet implemented is not yet implemented: file Foundation/FileManager.swift, line 430
fish: “swift run testmove” terminated by signal SIGILL (Illegal instruction)
```

Then I tried making it use `/var/cache` and it failed because of something different:

```console
felix@felix-X550LD ~/D/P/A/testmove> swift run testmove
Compile Swift Module 'SDGPersistence' (10 sources)
Linking ./.build/x86_64-unknown-linux/debug/testmove
Creating temporary file...
Creation succeeded.
Moving to cache...
The operation could not be completed
```

**Comment by SDGGiesbrecht** [comment]
Thank you for all your help, @felix91gr.
I am running out of ideas...

Permissions look fine. That *shouldn’t* be it.

The failure of the arbitrarily chosen directory is even weirder.

It will look into what `Foundation` does...

**Comment by SDGGiesbrecht** [comment]
The only way of getting the “Cross‐device moves...” message is if `Foundation` receives an `EXDEV` error from `Glibc`’s `rename(_:_:)`. ([`Foundation`’s source which catches the error](https://github.com/apple/swift-corelibs-foundation/blob/c2a41594235cc21d6a623637467f0db4b3c46cf9/Foundation/FileManager.swift#L480-L483) has gone unchanged for two years.)

According to the [GNU documentation for `rename(_:_:)`](http://www.gnu.org/software/libc/manual/html_node/Renaming-Files.html):
> EXDEV
> The two file names `newname` and `oldname` are on different file systems.

...and [for `EXDEV` itself](https://www.gnu.org/software/libc/manual/html_node/Error-Codes.html):

> Macro: *int* **EXDEV**
> “Invalid cross-device link.” An attempt to make an improper link across file systems was detected. This happens not only when you use `link` (see Hard Links) but also when you rename a file with `rename` (see Renaming Files).

But you (@felix91gr) said,
> This is on just one drive.

???

***

The `Foundation` source contains this note:
> // TODO: Copy and delete.

What happens if you use the `testmove` script from the earlier comment, but you adjust the `move` function instead (Line 145). Switch it to this:

```swift
try copy(source, to: destination)
try removeItem(at: source)
```

If it works we can try using `Glibc` directly, catch `EXDEV` if it occurs, and then do the above workaround instead of crashing.

**Comment by SDGGiesbrecht** [comment]
@felix91gr,

The workaround from the last comment is now live in `master`. Running `swift test` should be sufficient to see if it fixed everything or only fixed part of the issue.

**Comment by felix91gr** [comment]
Awesome! I'll try running it when I get back home :)

**Comment by felix91gr** [comment]
Allright. Here's what I found out today.

The failing test needs super user (`su` or `sudo`) access to pass.

### Setup

1. Downloaded a clean copy of the repo, 
2. Didn't change or checked out anything (used `master` directly),
3. Ran the command

I used 2 different commands.

### With `swift test`

This is the end of the console output:

```console
Test Case 'InternalTests.testVersionSelection' started at 2018-04-23 02:56:00.793

$ swift package init --type executable
Creating executable package: tool
Creating Package.swift
Creating README.md
Creating .gitignore
Creating Sources/
Creating Sources/tool/main.swift
Creating Tests/


$ git init
Initialized empty Git repository in /tmp/ca.solideogloria.SDGCommandLine.Tests.debug/tool/.git/


$ git add .


$ git commit --m Initialized.
[master (root-commit) 7d5c22d] Initialized.
 4 files changed, 27 insertions(+)
 create mode 100644 .gitignore
 create mode 100644 Package.swift
 create mode 100644 README.md
 create mode 100644 Sources/tool/main.swift


$ git add .


$ git commit --m "Version 1.0.0"
[master 56a8ed5] Version 1.0.0
 1 file changed, 1 insertion(+), 1 deletion(-)


$ git tag 1.0.0


$ git clone "file:///tmp/ca.solideogloria.SDGCommandLine.Tests.debug/tool" "/tmp/ca.solideogloria.SDGCommandLine.Tests.debug/Package Clones/tool" --branch 1.0.0 --depth 1 --config advice.detachedHead=false
Cloning into '/tmp/ca.solideogloria.SDGCommandLine.Tests.debug/Package Clones/tool'...


$ swift build --configuration release --static-swift-stdlib
Compile Swift Module 'tool' (1 sources)
Linking ./.build/x86_64-unknown-linux/release/tool


$ "/home/felix/.cache/ca.solideogloria.SDGCommandLine.Tests.debug/Versions/1.0.0/tool" some‐invalid‐argument another‐invalid‐argument
/bin/sh: 1: /home/felix/.cache/ca.solideogloria.SDGCommandLine.Tests.debug/Versions/1.0.0/tool: Permission denied
/home/felix/Documents/Programming/AwesomeSystemSwift/sdgCommandLine/Tests/SDGCommandLineTests/InternalTests.swift:201: error: InternalTests.testVersionSelection : failed - /bin/sh: 1: /home/felix/.cache/ca.solideogloria.SDGCommandLine.Tests.debug/Versions/1.0.0/tool: Permission denied
Test Case 'InternalTests.testVersionSelection' failed (7.018 seconds)
Test Case 'InternalTests.testVersionSubcommand' started at 2018-04-23 02:56:07.811
1.2.3
Test Case 'InternalTests.testVersionSubcommand' passed (0.0 seconds)
Test Case 'InternalTests.testXcode' started at 2018-04-23 02:56:07.812
Test Case 'InternalTests.testXcode' passed (0.0 seconds)
Test Suite 'InternalTests' failed at 2018-04-23 02:56:07.812
	 Executed 10 tests, with 1 failure (0 unexpected) in 15.202 (15.202) seconds
Test Suite 'debug.xctest' failed at 2018-04-23 02:56:07.812
	 Executed 20 tests, with 1 failure (0 unexpected) in 17.648 (17.648) seconds
Test Suite 'All tests' failed at 2018-04-23 02:56:07.812
	 Executed 20 tests, with 1 failure (0 unexpected) in 17.648 (17.648) seconds
```

Same as before, I think.

### With `sudo swift test`

```console
Test Case 'InternalTests.testVersionSelection' started at 2018-04-23 02:57:13.228

$ swift package init --type executable
Creating executable package: tool
Creating Package.swift
Creating README.md
Creating .gitignore
Creating Sources/
Creating Sources/tool/main.swift
Creating Tests/


$ git init
Initialized empty Git repository in /tmp/ca.solideogloria.SDGCommandLine.Tests.debug/tool/.git/


$ git add .


$ git commit --m Initialized.
[master (root-commit) 2427402] Initialized.
 4 files changed, 27 insertions(+)
 create mode 100644 .gitignore
 create mode 100644 Package.swift
 create mode 100644 README.md
 create mode 100644 Sources/tool/main.swift


$ git add .


$ git commit --m "Version 1.0.0"
[master 631dbc8] Version 1.0.0
 1 file changed, 1 insertion(+), 1 deletion(-)


$ git tag 1.0.0


$ git clone "file:///tmp/ca.solideogloria.SDGCommandLine.Tests.debug/tool" "/tmp/ca.solideogloria.SDGCommandLine.Tests.debug/Package Clones/tool" --branch 1.0.0 --depth 1 --config advice.detachedHead=false
Cloning into '/tmp/ca.solideogloria.SDGCommandLine.Tests.debug/Package Clones/tool'...


$ swift build --configuration release --static-swift-stdlib
Compile Swift Module 'tool' (1 sources)
Linking ./.build/x86_64-unknown-linux/release/tool


$ "/root/.cache/ca.solideogloria.SDGCommandLine.Tests.debug/Versions/1.0.0/tool" some‐invalid‐argument another‐invalid‐argument
some‐invalid‐argument another‐invalid‐argument


$ "/root/.cache/ca.solideogloria.SDGCommandLine.Tests.debug/Versions/1.0.0/tool" some‐invalid‐argument another‐invalid‐argument
some‐invalid‐argument another‐invalid‐argument


$ git ls-remote "file:///tmp/ca.solideogloria.SDGCommandLine.Tests.debug/tool" master
631dbc86fe9a8be05cafa5d00c9a25de77294432	refs/heads/master


$ git clone "file:///tmp/ca.solideogloria.SDGCommandLine.Tests.debug/tool" "/tmp/ca.solideogloria.SDGCommandLine.Tests.debug/Package Clones/tool" --depth 1 --config advice.detachedHead=false
Cloning into '/tmp/ca.solideogloria.SDGCommandLine.Tests.debug/Package Clones/tool'...


$ swift build --configuration release --static-swift-stdlib
Compile Swift Module 'tool' (1 sources)
Linking ./.build/x86_64-unknown-linux/release/tool


$ "/root/.cache/ca.solideogloria.SDGCommandLine.Tests.debug/Versions/Development/631dbc86fe9a8be05cafa5d00c9a25de77294432/tool" some‐invalid‐argument another‐invalid‐argument
some‐invalid‐argument another‐invalid‐argument


$ git ls-remote "file:///tmp/ca.solideogloria.SDGCommandLine.Tests.debug/tool" master
631dbc86fe9a8be05cafa5d00c9a25de77294432	refs/heads/master


$ "/root/.cache/ca.solideogloria.SDGCommandLine.Tests.debug/Versions/Development/631dbc86fe9a8be05cafa5d00c9a25de77294432/tool" some‐invalid‐argument another‐invalid‐argument
some‐invalid‐argument another‐invalid‐argument


$ git clone "file:///tmp/ca.solideogloria.SDGCommandLine.Tests.debug/tool" "/tmp/ca.solideogloria.SDGCommandLine.Tests.debug/Package Clones/tool" --branch 1.0.0 --depth 1 --config advice.detachedHead=false
Cloning into '/tmp/ca.solideogloria.SDGCommandLine.Tests.debug/Package Clones/tool'...


$ swift build --configuration release --static-swift-stdlib
Compile Swift Module 'tool' (1 sources)
Linking ./.build/x86_64-unknown-linux/release/tool


$ "/tmp/ca.solideogloria.SDGCommandLine.Tests.debug/6BC37558-C1CE-482E-88B8-8441CB89BECF/tool"


Test Case 'InternalTests.testVersionSelection' passed (4.183 seconds)
Test Case 'InternalTests.testVersionSubcommand' started at 2018-04-23 02:57:17.412
1.2.3
Test Case 'InternalTests.testVersionSubcommand' passed (0.0 seconds)
Test Case 'InternalTests.testXcode' started at 2018-04-23 02:57:17.412
Test Case 'InternalTests.testXcode' passed (0.0 seconds)
Test Suite 'InternalTests' passed at 2018-04-23 02:57:17.412
	 Executed 10 tests, with 0 failures (0 unexpected) in 8.458 (8.458) seconds
Test Suite 'debug.xctest' passed at 2018-04-23 02:57:17.412
	 Executed 20 tests, with 0 failures (0 unexpected) in 9.772 (9.772) seconds
Test Suite 'All tests' passed at 2018-04-23 02:57:17.412
	 Executed 20 tests, with 0 failures (0 unexpected) in 9.772 (9.772) seconds
```

It works! Which is pretty cool.

There's probably some reason behind this. Why does it need super user access? I will read the output in more detail, it's probably got something revealing.

**Comment by SDGGiesbrecht** [comment]
> Same as before, I think.

Not really. Here the move succeeded. The permissions complaint came later when it tries to execute the version it put there.

> Why does it need super user access?

No idea. It is not needed on my machine.

You apparently do not have permission to execute `/home/felix/.cache/ca.solideogloria.SDGCommandLine.Tests.debug/Versions/1.0.0/tool`. The directories from `ca.soli...` onward are created by `Foundation`’s `createDirectory(at:withIntermediateDirectories:attributes:)` (with `attributes: nil`), and `tool` is what was moved there (via copy when necessary). All of it is wiped out between tests.

**Comment by SDGGiesbrecht** [comment]
@felix91gr,

> Will SDGCommandLine still be usable on my machine?

If you build a tool which links against SDGCommandLine to provide its interface, most of the resulting tool will work normally:

```shell
$ my‐tool do‐something
Nothing wrong. Executing normally...
```

The only glitch is that the provided `•use‐version` option would require `sudo` (which, when I think about it, is not really a bad thing.) From its `help` description:

```
•use‐version [version] Attempts to download and temporarily use the specified version insead of one which is installed.
```

The user would see approximately this:

```shell
$ my‐tool do‐something •use‐version 1.0.0
Downloading...
Building...
Lauching...
Permission denied.

$ sudo my‐tool do‐something •use‐version 1.0.0
Downloading...
Building...
Launching...
Nothing wrong. Executing normally...
```

**Comment by felix91gr** [comment]
Well, that's fine by me. Thanks Jeremy, for your patience ^^

This week I'm expecting to merge your entries to the index.

**Comment by SDGGiesbrecht** [comment]
I like to leave issues open until it the fix is not just merged, but released. That way the issue is easier to find and it is less likely for duplicates to be reported.

**Comment by SDGGiesbrecht** [comment]
The general crash is fixed in version 0.3.0.

(Permissions may still require `sudo` in some situations.)
